lib/components/Forms/Fields/NumberInput/index.tsx

```diff
@@ -0,0 +1,77 @@
+import { forwardRef, useEffect, useState } from "react"
+import { Input, InputProps } from "../Input"
+
+const FORMAT = /^([0-9]+)$/
+// eslint-disable-next-line no-useless-escape
+const FORMAT_WITH_DECIMALS = /^([0-9]+)(?:[\.,]([0-9]+)?)?$/
+
+const parseValue = (value: string) => parseFloat(value.replace(",", "."))
+
+type NumberInputProps = Omit<InputProps, "value" | "type" | "onChange"> & {
+  locale: string
+  value?: number
+  maxDecimals?: number
+  onChange?: (value: number | null) => void
+}
+
+export const NumberInput = forwardRef<HTMLInputElement, NumberInputProps>(
+  ({ locale, value, maxDecimals, onChange, ...props }, ref) => {
+    const formatValue = (value: number, maxDecimals?: number) =>
+      new Intl.NumberFormat(locale, {
+        maximumFractionDigits: maxDecimals,
+        useGrouping: false,
+      }).format(value)
+
+    const [fieldValue, setFieldValue] = useState<string>(() =>
+      value != null ? formatValue(value, maxDecimals) : ""
+    )
+
+    const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+      const value = event.target.value
+
+      if (!value) {
+        setFieldValue("")
+        return onChange?.(null)
+      }
+
+      const match = value.match(
+        maxDecimals !== 0 ? FORMAT_WITH_DECIMALS : FORMAT
+      )
+      if (!match) return
+
+      const [_, integers, decimals] = match
+      if (maxDecimals && decimals && decimals.length > maxDecimals) return
+
+      const valueWithNoLeadingZero = value.match(/^0\d$/)
+        ? value.slice(1)
+        : value
+
+      const valueAsNumber = parseFloat(`${integers}.${decimals ?? 0}`)
+
+      setFieldValue(valueWithNoLeadingZero)
+      onChange?.(valueAsNumber)
+    }
+
+    // We need the logic "parseValue(fieldValue) !== value" below to ensure we don't remove the decimal separator while the user is modifying the decimals.
+    // For example, if the input value is 18,3 and I want to change it to 18,5. I'll delete the last digit and the input will become "18,"
+    // At this point, we have fired the "onChange" callback with the value 18 because it's the most precise valid value ("18," isn't a valid one),
+    // then the useEffect below will be triggered, since the "value" changed from "18,3" to "18", and it'll not modify the "fieldValue" to keep the decimal separator
+    // unless if the value really changed.
+    useEffect(() => {
+      if (value === null) return setFieldValue("")
+      if (value != null && parseValue(fieldValue) !== value)
+        setFieldValue(formatValue(value, maxDecimals))
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [fieldValue, value])
```

Comment:
Can we change this for a const? I think is more performance, because useEffect requieres extra renders

```suggestion
const fieldValue = !value || parseValue(value) !== value ? formatValue(value) : value
```

Reply:
We must have both a state (`fieldValue`) and the `useEffect` inside this component to handle [this scenario that we already discussed in Gamma's PR](https://github.com/factorialco/factorial/pull/47179#discussion_r1695697766). If we don't have another state, it makes impossible to keep the decimal separator while the user is typing.

Reply:
But if the user type a , and this has no sence (because in the local defined should be a .) makes sense to replace it while typing no?

Reply:
Yeah I remember... The problem is the input is a number, then you loose that information (the decimal point)... ok ok

Reply:
> But if the user type a , and this has no sence (because in the local defined should be a .) makes sense to replace it while typing no?

It depends if we want to force to use the same separator defined in the locale or not. Right now we don't restrict that, so we allow both separators independently of the user locale. I've implemented it like this because that's the same behavior of the native browser's `input type="number"`.

Reply:
Then if I'm in Spanish can I type 1.223,43?

Reply:
Not really because we don't support thousand separators for now. to keep the same behavior of the native input type number